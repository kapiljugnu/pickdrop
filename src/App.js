import React, { Component } from 'react';
import { withScriptjs } from 'react-google-maps';
import Map from './Map';
import LocationInput from './Location/Input';
import './App.css';
import { getPath } from './Api/action';

export class App extends Component {

  state = { path: [] }

  onLocationSubmit = (pick, drop) => {
    if (pick && drop) {
      getPath({ pick, drop }).then(({ data: { path = [] } }) => this.setState({ path }));
    }
  }

  render() {
    return (
      <div className="container">
        <div className="form">
          <LocationInput
            onSubmit={this.onLocationSubmit} />
        </div>
        <Map
          path={this.state.path}
          containerElement={<div className="map" />}
          mapElement={<div style={{ height: `100%` }} />}
        />
      </div>
    );
  }
}

export default withScriptjs(App);