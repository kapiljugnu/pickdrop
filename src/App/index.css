.container {
    display: flex;
    flex-direction: row;
    position: relative;
    min-height: 100vh;
}

.form {
    flex: 1;
    order: 1;
    margin: 0;
    padding: 0;
}

.map{
    flex: 3;
    order: 2;
    position: relative;
    min-height: 100%;
}