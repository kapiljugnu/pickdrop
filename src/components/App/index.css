.container {
    display: flex;
    flex-direction: row;
    position: relative;
    min-height: 100vh;
}

.form {
    flex: 1;
    order: 1;
    margin: 0;
    padding: 0;
}

.map{
    flex: 3;
    order: 2;
    position: relative;
    min-height: 100%;
}

.error {
    color: red;
    margin: 12px 0;
}

.loader {
    border: 5px solid #f3f3f3;
    border-top: 5px solid #555;
    border-radius: 50%;
    width: 50px;
    height: 50px;
    animation: spin 2s linear infinite;
  }

.form .loader {
  margin-top: 3px;
  width:20px;
  height: 20px;
 }

  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }

    100% {
      transform: rotate(360deg);
    }
  }