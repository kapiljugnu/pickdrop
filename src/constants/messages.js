export const SERVER_ERROR = 'Error occured while fetching the data, please try again.'
export const INPUT_ERROR = 'Please provide starting location and drop-off location and submit again.'