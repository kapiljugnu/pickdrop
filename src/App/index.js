import React, { Component } from 'react';
import { withScriptjs } from 'react-google-maps';
import Map from '../Map';
import LocationInput from '../Location/Input';
import './index.css';
import { getPath } from '../Api/action';
import Direction from '../Map/Direction';
import { SERVER_ERROR, INPUT_ERROR } from '../constants/messages';

export class App extends Component {

  state = { path: [], hasError: false, message: '' }

  onLocationSubmit = (pick, drop) => {
    this.setState({ hasError: false, message: '' })
    if (pick && drop) {
      getPath({ pick, drop })
        .then(({ data: { path = [] } }) => this.setState({ path }))
        .catch((error) => { this.setState({ hasError: true, message: SERVER_ERROR }) });
    } else {
      this.setState({ hasError: true, message: INPUT_ERROR })
    }
  }

  onReset = () => {
    this.setState({ path: [], hasError: false, message: '' });
  }

  render() {
    return (
      <Direction path={this.state.path}>
        {
          ({ directions, distance, duration }) => <div className="container">
            <div className="form">
              <LocationInput distance={distance.toString()} duration={duration.toString()} onSubmit={this.onLocationSubmit} onReset={this.onReset} />
              {this.state.hasError && <div className="error">{this.state.message}</div>}
            </div>
            <Map
              directions={directions}
              containerElement={<div className="map" />}
              mapElement={<div style={{ height: `100%` }} />}
            />
          </div>
        }
      </Direction>
    );
  }
}

export default withScriptjs(App);