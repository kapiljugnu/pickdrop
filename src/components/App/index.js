import React, { Component } from 'react';
import { withScriptjs } from 'react-google-maps';
import Map from '../Map';
import Location from '../Location';
import './index.css';
import getPath from '../../services/getPath';
import Direction from '../Map/Direction';
import { SERVER_ERROR, INPUT_ERROR } from '../../constants/messages';

export class App extends Component {

  state = { path: [], hasError: false, message: '', isFetching: false }

  onLocationSubmit = (pick, drop) => {
    if (!pick || !drop) {
      this.setState({ hasError: true, message: INPUT_ERROR, isFetching: false });
      return;
    }
    this.setState({ hasError: false, message: '', isFetching: true })
    getPath({ pick, drop })
      .then(({ data: { path = [] } }) => this.setState({ path }))
      .catch((error) => { this.setState({ hasError: true, message: SERVER_ERROR }) })
      .finally(() => { this.setState({ isFetching: false }) })
  }

  onReset = () => {
    this.setState({ path: [], hasError: false, message: '' });
  }

  render() {
    return (
      <Direction path={this.state.path}>
        {
          ({ directions, distance, duration }) => <div className="container">
            <div className="form">
              <Location distance={distance.toString()} duration={duration.toString()} onSubmit={this.onLocationSubmit} onReset={this.onReset} />
              {this.state.isFetching && <div className="loader" />}
              {this.state.hasError && <div className="error">{this.state.message}</div>}
            </div>
            <Map
              directions={directions}
              containerElement={<div className="map" />}
              mapElement={<div style={{ height: `100%` }} />}
            />
          </div>
        }
      </Direction>
    );
  }
}

export default withScriptjs(App);