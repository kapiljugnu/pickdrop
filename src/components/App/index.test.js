import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App, { App as ScriptChild } from './index';
import { SERVER_ERROR, INPUT_ERROR } from '../../constants/messages';


describe('App.js', () => {

  const loadingElement = <div style={{ height: `100%` }} />
  const googleMapURL = 'googleMapURL';

  it('renders without crashing', () => {
    const app = mount(<App
      loadingElement={loadingElement}
      googleMapURL={googleMapURL}
    />).setState({ loadingState: 'LOADED' });
    expect(toJson(app)).toMatchSnapshot();
  });

  it('onReset the app', () => {
    const appTree = mount(<App loadingElement={loadingElement}
      googleMapURL={googleMapURL} />).setState({ loadingState: 'LOADED' });
    const app = appTree.find(ScriptChild).setState({ hasError: true, message: 'mock' });
    expect(app.state().message).toBe('mock');
    expect(app.state().hasError).toBeTruthy();
    app.instance().onReset();
    expect(app.state().hasError).toBeFalsy();
    expect(app.state().message).toBe('');
  });

  it('onSubmit the location with no parameter', () => {
    const appTree = mount(<App loadingElement={loadingElement}
      googleMapURL={googleMapURL} />).setState({ loadingState: 'LOADED' });
    const app = appTree.find(ScriptChild);
    app.instance().onLocationSubmit();
    expect(app.state().hasError).toBeTruthy();
    expect(app.state().message).toBe(INPUT_ERROR);
  });

  it('onSubmit the location with parameter', () => {
    const appTree = mount(<App loadingElement={loadingElement}
      googleMapURL={googleMapURL} />).setState({ loadingState: 'LOADED' });
    const app = appTree.find(ScriptChild);
    app.instance().onLocationSubmit('location pick', 'location drop');
    expect(app.state().hasError).toBeFalsy();
    expect(app.state().message).toBe('');
    expect(app.state().isFetching).toBeTruthy();
    process.nextTick(() => {
      expect(app.state().isFetching).toBeFalsy();
      expect(app.state().path).toEqual([1, 2])
    });
  });
});
