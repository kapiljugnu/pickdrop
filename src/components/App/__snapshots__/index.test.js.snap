// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<withScriptjs(App)
  googleMapURL="googleMapURL"
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
>
  <App>
    <Direction
      path={Array []}
    >
      <div
        className="container"
      >
        <div
          className="form"
        >
          <Input
            distance=""
            duration=""
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <form>
              <label
                htmlFor="pick"
              >
                Starting Location
              </label>
              <StandaloneSearchBox
                onPlacesChanged={[Function]}
              >
                <div
                  className="locationInput"
                >
                  <input
                    id="pick"
                    placeholder="Starting Location"
                    required={true}
                    type="text"
                  />
                  <i
                    onClick={[Function]}
                  >
                    <_default>
                      <svg
                        height="24"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M24 20.188l-8.315-8.209 8.2-8.282-3.697-3.697-8.212 8.318-8.31-8.203-3.666 3.666 8.321 8.24-8.206 8.313 3.666 3.666 8.237-8.318 8.285 8.203z"
                        />
                      </svg>
                    </_default>
                  </i>
                </div>
              </StandaloneSearchBox>
              <label
                htmlFor="drop"
              >
                Drop-off point
              </label>
              <StandaloneSearchBox
                onPlacesChanged={[Function]}
              >
                <div
                  className="locationInput"
                >
                  <input
                    id="drop"
                    placeholder="Drop-off point"
                    required={true}
                    type="text"
                  />
                  <i
                    onClick={[Function]}
                  >
                    <_default>
                      <svg
                        height="24"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M24 20.188l-8.315-8.209 8.2-8.282-3.697-3.697-8.212 8.318-8.31-8.203-3.666 3.666 8.321 8.24-8.206 8.313 3.666 3.666 8.237-8.318 8.285 8.203z"
                        />
                      </svg>
                    </_default>
                  </i>
                </div>
              </StandaloneSearchBox>
              <br />
              <input
                onClick={[Function]}
                type="submit"
                value="Submit"
              />
              <input
                onClick={[Function]}
                type="button"
                value="Reset"
              />
            </form>
          </Input>
        </div>
        <withGoogleMap(Map)
          containerElement={
            <div
              className="map"
            />
          }
          directions={null}
          mapElement={
            <div
              style={
                Object {
                  "height": "100%",
                }
              }
            />
          }
        >
          <div
            className="map"
          >
            <div
              style={
                Object {
                  "height": "100%",
                }
              }
            />
            <div />
          </div>
        </withGoogleMap(Map)>
      </div>
    </Direction>
  </App>
</withScriptjs(App)>
`;
